### PR TITLE
WIP: docs for writing an apiserver

### PR DESCRIPTION
**Pull request by MHBauer**
Writing down what I have learned while writing the service-catalog apiserver in a guide format.

### CODE REVIEW

**Comment by eparis** [comment]
@deads2k

File: contributors/devel/writing-an-apiserver.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+               <versions>
+               v1alpha1/
+               validation/
+      apiserver/
+      cmd/server/
+      registry/
+          <api-group-name>/
+              rest/
+                  <storage>
+              <individual types>/
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: s/types/resources/

File: contributors/devel/writing-an-apiserver.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                  storage.go
+                  strategy.go
+```
+
+These structures follow the structure of kubernetes.
+
+## cmd/
+
+This toplevel directory contains the package that defines the output
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: top-level

File: contributors/devel/writing-an-apiserver.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+resources, and should not be serialized upwards to a client. These
+unversioned resources must be able to hold any version, as all
+versioned resources will be converted into their unversioned form
+before being persisted.
+
+The *versioned* `types.go` resides in a directory with the version
+name that follows
+[kubernetes conventions](api_changes.md#alpha-beta-and-stable-versions). The
+defined structs in this file should have `json` annotations.
+
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: Might want to link to API conventions and adding an API group here.

File: contributors/devel/writing-an-apiserver.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+1. the lifetime of your API is separrate from that of the core APIs.
+1. Doing something experimental and need access to the internal APIs.
+1. Want to federate your API to the kubernetes core APIs so it appears
+   as a first class API from an external user of kubernetes.
+
+
+
+# Glossary
+
+`Upwards` means towards a client such as kubectl. `Downwards` means
```
Comment (MHBauer) [inline contributors/devel/writing-an-apiserver.md]: What is the other downward API?

I'll figure something out.

**Comment by MHBauer** [comment]
I don't think the header/subheader relationship is clear in the rendered docs, so I will move to typing out the full path.

File: contributors/devel/writing-an-apiserver.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
+Writing an API Server
+=====================
+
+# Why write an apiserver?
+
+Reasons to write an apiserver:
+
+1. You want a new API for kubernetes, but it is unlikely to be merged
+   into core APIs.
+1. the lifetime of your API is separrate from that of the core APIs.
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: separate

File: contributors/devel/writing-an-apiserver.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+1. Doing something experimental and need access to the internal APIs.
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: 'You're doing'

File: contributors/devel/writing-an-apiserver.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+1. Want to federate your API to the kubernetes core APIs so it appears
```
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: 'You want'

File: contributors/devel/writing-an-apiserver.md (hunk 3 above)
Comment (pmorie) [inline contributors/devel/writing-an-apiserver.md]: Nevermind, I see you've done it above.

File: contributors/devel/writing-an-apiserver.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+### `apis/`
+
+This contains all of the type definitions along with conversion
+between versions, validation of types, and the
+registration/installation of the API-Group.
+
+#### `pkg/apis/<API-Group>/`
+
+Each API-Group has it's own named directory, which should be named by
+running all the words in it together with no spaces or dashes.
```
Comment (vaikas-google) [inline contributors/devel/writing-an-apiserver.md]: perhaps add lowercased too?

**Comment by philips** [comment]
@MHBauer what do you want to do with this?

**Comment by pmorie** [comment]
I think literally any doc is an improvement, but I don't expect this PR to fully solve the problem of adding a comprehensive doc.  IMO the folks developing API machinery ought to develop the doc and keep it up to date.

Suggestion: finish adding something good from this PR, doesn't have to be perfect, next, integrate making the docs comprehensive and maintaining said docs part of the API machinery development process.

**Comment by pmorie** [comment]
cc @vaikas-google and @pwittrock since we've been talking about this problem

@MHBauer do you think you will have some bandwidth to get this to a point where it can be checked in and built upon soon?  If not, I am happy to pick it up.

**Pull request closed without merging**